### ISSUE TITLE
Some containers don't start with Engine 23.0

### ISSUE THREAD

**Issue by rrajkomar** [moby/moby#44928]
### Description

After upgrading my docker-ce to 23.0, I started having some containers that would no longer start with the following message : 
```
failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "/usr/local/bin/entrypoint.sh": stat [my_entrypoint_sh_absolute_filepath]: no such file or directory: unknown
```

Do note that even with an entrypoint set to `/bin/sh` the same issue happens

I also has the same error while rebuillding a python based container image : 
```
runc run failed: unable to start container process: exec: "/bin/sh": stat /bin/sh: no such file or directory
```

### Reproduce

apt update
apt upgrade
to install docker version 23.

### Expected behavior

Containers should start as they did with 20.10

### docker version

```bash
Client: Docker Engine - Community
 Version:           23.0.0
 API version:       1.42
 Go version:        go1.19.5
 Git commit:        e92dd87
 Built:             Wed Feb  1 17:43:19 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.0
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.5
  Git commit:       d7573ab
  Built:            Wed Feb  1 17:43:19 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.16
  GitCommit:        31aa4358a36870b21a992d3ad2bef29e1d693bec
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.15.1
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 175
  Running: 132
  Paused: 0
  Stopped: 43
 Images: 814
 Server Version: 23.0.0
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 31aa4358a36870b21a992d3ad2bef29e1d693bec
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
 Kernel Version: 5.10.0-0.bpo.12-amd64
 Operating System: Debian GNU/Linux 10 (buster)
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 62.63GiB
 Name: silvercity
 ID: TPZ4:ABIL:UYYM:3HTN:K564:RF7X:YVKD:D6TF:LOYY:NIYX:A2M4:IFV5
 Docker Root Dir: /opt/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by neersighted** [comment]
There's not enough here to go off of -- a true reproduction is probably needed, but daemon logs may also help.

* Can you provide logs from docker.service? `journalctl -u docker.service`
* Can you provide a reproduction that works anywhere, e.g. Docker-in-Docker? An example image + run line that reproduces the failure on other machines would be the ideal form.

**Comment by rrajkomar** [comment]
Sorry it took me so long to reply, I couldn't switch back to the 23.0 Engine as I needed my containers to be operational.
I just attempted again to run my containers with the new engin while monitoring the docker logs.

It looks like the issue is related to one of the entrypoint.sh related commands (most likely the copy since the container start process can't seem to find the file)

Here's some basic setup that should allow to get the issue. Here's what I have when I try building from it.
[test.tar.gz](https://github.com/moby/moby/files/10673008/test.tar.gz)

```
[+] Building 0.4s (6/7)                                                                                                                                                                                            
 => [internal] load build definition from Dockerfile                                                                                                                                                          0.0s
 => => transferring dockerfile: 207B                                                                                                                                                                          0.0s
 => [internal] load .dockerignore                                                                                                                                                                             0.0s
 => => transferring context: 2B                                                                                                                                                                               0.0s
 => [internal] load metadata for docker.io/library/php:8.1.12-fpm-alpine3.16                                                                                                                                  0.3s
 => [internal] load build context                                                                                                                                                                             0.0s
 => => transferring context: 2B                                                                                                                                                                               0.0s
 => CACHED [1/3] FROM docker.io/library/php:8.1.12-fpm-alpine3.16@sha256:fb54571d659a7e474377a2821a7a874667da4db48e9b9062ff088a8ebd7f3cea                                                                     0.0s
 => ERROR [2/3] COPY ./entrypoint.sh /usr/local/bin/entrypoint.sh                                                                                                                                             0.0s
------
 > [2/3] COPY ./entrypoint.sh /usr/local/bin/entrypoint.sh:
------
Dockerfile:3
--------------------
   1 |     FROM php:8.1.12-fpm-alpine3.16
   2 |     
   3 | >>> COPY ./entrypoint.sh /usr/local/bin/entrypoint.sh
   4 |     RUN chmod +x /usr/local/bin/entrypoint.sh
   5 |     ENTRYPOINT ["/usr/local/bin/entrypoint.sh"]
--------------------
ERROR: failed to solve: failed to compute cache key: failed to calculate checksum of ref moby::h0n6f23r6pziqlztg95mm2gki: "/entrypoint.sh": not found
```

Note that the exact same configuration works fine with the previous version of the engine (20.10) :  
```
Sending build context to Docker daemon  3.072kB
Step 1/4 : FROM php:8.1.12-fpm-alpine3.16
8.1.12-fpm-alpine3.16: Pulling from library/php
ca7dd9ec2225: Already exists 
1b78b4fe0ca1: Already exists 
9d6040f2a28f: Already exists 
0e2e66b89284: Already exists 
c91bb44404d4: Already exists 
568a7ba86886: Already exists 
681d08e03e4f: Already exists 
dafedc9ad152: Already exists 
0449b3a1644b: Already exists 
45c6b0057562: Already exists 
Digest: sha256:fb54571d659a7e474377a2821a7a874667da4db48e9b9062ff088a8ebd7f3cea
Status: Downloaded newer image for php:8.1.12-fpm-alpine3.16
 ---> 98cbdd4fe903
Step 2/4 : COPY ./entrypoint.sh /usr/local/bin/entrypoint.sh
 ---> 125145b93893
Step 3/4 : RUN chmod +x /usr/local/bin/entrypoint.sh
 ---> Running in 865cd9d6eb22
Removing intermediate container 865cd9d6eb22
 ---> f6cea353a52d
Step 4/4 : ENTRYPOINT ["/usr/local/bin/entrypoint.sh"]
 ---> Running in 83959d53efdf
Removing intermediate container 83959d53efdf
 ---> 96e960830ec9
Successfully built 96e960830ec9
```


Here the relevant log output from trying to create all my containers:  
```
Feb  7 09:04:36 silvercity dhclient[30375]: DHCPDISCOVER on docker0 to 255.255.255.255 port 67 interval 13
Feb  7 09:04:49 silvercity dhclient[30375]: DHCPDISCOVER on docker0 to 255.255.255.255 port 67 interval 14
Feb  7 09:04:49 silvercity dockerd[21179]: time="2023-02-07T09:04:49.915304507+01:00" level=error msg="stream copy error: reading from a closed fifo"
Feb  7 09:04:49 silvercity dockerd[21179]: time="2023-02-07T09:04:49.915305446+01:00" level=error msg="stream copy error: reading from a closed fifo"
Feb  7 09:04:50 silvercity dockerd[21179]: time="2023-02-07T09:04:50.030495038+01:00" level=error msg="stream copy error: reading from a closed fifo"
Feb  7 09:04:50 silvercity dockerd[21179]: time="2023-02-07T09:04:50.030513495+01:00" level=error msg="stream copy error: reading from a closed fifo"
Feb  7 09:04:50 silvercity dockerd[21179]: time="2023-02-07T09:04:50.535085969+01:00" level=error msg="493ba9f13387d23de468116f54a995038717feb18926776e065b548c9b00017f cleanup: failed to delete container from containerd: no such container"
Feb  7 09:04:50 silvercity systemd[4769]: run-docker-netns-4042cee4acd8.mount: Succeeded.
Feb  7 09:04:50 silvercity systemd[4769]: opt-docker-overlay2-9ee4598d221e635112216b09273228bbceaeee8566501d17b03536d419a3386e-merged.mount: Succeeded.
Feb  7 09:04:50 silvercity systemd[1]: run-docker-netns-4042cee4acd8.mount: Succeeded.
Feb  7 09:04:50 silvercity systemd[1]: opt-docker-overlay2-9ee4598d221e635112216b09273228bbceaeee8566501d17b03536d419a3386e-merged.mount: Succeeded.
Feb  7 09:04:51 silvercity dockerd[21179]: time="2023-02-07T09:04:51.069210630+01:00" level=error msg="stream copy error: reading from a closed fifo"
Feb  7 09:04:51 silvercity dockerd[21179]: time="2023-02-07T09:04:51.069234613+01:00" level=error msg="stream copy error: reading from a closed fifo"
Feb  7 09:04:51 silvercity dockerd[21179]: time="2023-02-07T09:04:51.083096591+01:00" level=error msg="689a8c3ab6abce9d68091811ee743dd05b5f758bf2d2f9f47a797f6ed4c97d55 cleanup: failed to delete container from containerd: no such container"
Feb  7 09:04:51 silvercity systemd[4769]: run-docker-netns-894d6570e192.mount: Succeeded.
Feb  7 09:04:51 silvercity systemd[4769]: opt-docker-overlay2-1a0d18a8d00b7a474d2f0fccaaa92344f305c3870456c595178cce60c9e7cdb5-merged.mount: Succeeded.
Feb  7 09:04:51 silvercity systemd[1]: run-docker-netns-894d6570e192.mount: Succeeded.
Feb  7 09:04:51 silvercity systemd[1]: opt-docker-overlay2-1a0d18a8d00b7a474d2f0fccaaa92344f305c3870456c595178cce60c9e7cdb5-merged.mount: Succeeded.
Feb  7 09:04:51 silvercity dockerd[21179]: time="2023-02-07T09:04:51.354876770+01:00" level=error msg="stream copy error: reading from a closed fifo"
Feb  7 09:04:51 silvercity dockerd[21179]: time="2023-02-07T09:04:51.354919155+01:00" level=error msg="stream copy error: reading from a closed fifo"
Feb  7 09:04:51 silvercity dockerd[21179]: time="2023-02-07T09:04:51.382766458+01:00" level=error msg="stream copy error: reading from a closed fifo"
Feb  7 09:04:51 silvercity dockerd[21179]: time="2023-02-07T09:04:51.382785090+01:00" level=error msg="stream copy error: reading from a closed fifo"
Feb  7 09:04:51 silvercity dockerd[21179]: time="2023-02-07T09:04:51.385959979+01:00" level=error msg="8907c502a79f2e6358f9116d561bb5820e06213a375d0f30f5210fa2562c4665 cleanup: failed to delete container from containerd: no such container"
Feb  7 09:04:51 silvercity systemd[4769]: run-docker-netns-c9786d8bd754.mount: Succeeded.
Feb  7 09:04:51 silvercity systemd[4769]: opt-docker-overlay2-225f0c9d49f9cedd21e5cd8da86f9f08d53e99b0b0ec6d0f8f512db0db818b2a-merged.mount: Succeeded.
Feb  7 09:04:51 silvercity systemd[1]: run-docker-netns-c9786d8bd754.mount: Succeeded.
Feb  7 09:04:51 silvercity systemd[1]: opt-docker-overlay2-225f0c9d49f9cedd21e5cd8da86f9f08d53e99b0b0ec6d0f8f512db0db818b2a-merged.mount: Succeeded.
Feb  7 09:04:51 silvercity systemd[4769]: run-docker-netns-403bdc31e62f.mount: Succeeded.
Feb  7 09:04:51 silvercity systemd[4769]: opt-docker-overlay2-a2c9fe4401de13b9c6c30f66d403f6a11460ae22e2a6a42a9dbfb056dda60c11-merged.mount: Succeeded.
Feb  7 09:04:51 silvercity systemd[1]: run-docker-netns-403bdc31e62f.mount: Succeeded.
Feb  7 09:04:51 silvercity systemd[1]: opt-docker-overlay2-a2c9fe4401de13b9c6c30f66d403f6a11460ae22e2a6a42a9dbfb056dda60c11-merged.mount: Succeeded.
Feb  7 09:04:51 silvercity dockerd[21179]: time="2023-02-07T09:04:51.886479600+01:00" level=error msg="97052ac4f0e2b6c702900ab23df110be1cfeb570868136511810382fbdd033ac cleanup: failed to delete container from containerd: no such container"
Feb  7 09:04:51 silvercity systemd[1]: run-docker-netns-366cbb21b10a.mount: Succeeded.
Feb  7 09:04:51 silvercity systemd[1]: opt-docker-overlay2-21099776cd855947ea7dc6ba979c7f3a386e1c0025cf585b751bf44198562e89-merged.mount: Succeeded.
Feb  7 09:04:51 silvercity dockerd[21179]: time="2023-02-07T09:04:51.970958869+01:00" level=error msg="9f6fd46769a5ff77e5a04baa9a196cabf414c1bc1493387f91cc9174b15cca19 cleanup: failed to delete container from containerd: no such container"
Feb  7 09:04:51 silvercity systemd[4769]: run-docker-netns-366cbb21b10a.mount: Succeeded.
Feb  7 09:04:51 silvercity systemd[4769]: opt-docker-overlay2-21099776cd855947ea7dc6ba979c7f3a386e1c0025cf585b751bf44198562e89-merged.mount: Succeeded.
```

Also from the moment this issue occurs, I can't seem to rebuild any image as I keep getting this issue :  
```
runc run failed: unable to start container process: exec: "/bin/sh": stat /bin/sh: no such file or directory
```

**Comment by thaJeztah** [comment]
- Possibly related; https://github.com/moby/moby/issues/44943

**Comment by neersighted** [comment]
I'm unable to reproduce by following your directions (build on 20.10, run on 23.0), with or without BuildKit. I think we're skipping over vital context; could you please provide a step-by-step reproduction?

Notably:
* You've introduced image building; what exact command are you using to build your image on each version?
* Can you provide an export of the image that fails with `/bin/sh` not found? `docker save` will do this for you.
* What is the run command used to start the containers that fail?
* Is there anything like AppArmor or SELinux in play? The buildkit build shows a 2B transfer for the build context, which seems to indicate that the entrypoint script was never transferred to the daemon.

**Comment by rrajkomar** [comment]
`by following your directions (build on 20.10, run on 23.0), with or without BuildKit.`
> I think you misunderstood : both build and run are attemped on the same version.
On 20.10 : everything is fine
On 23.0 : the build looks fine but upon starting the containers, I get the entrypoint.sh not found error.

`You've introduced image building; what exact command are you using to build your image on each version?`
> Beside the provided instructions, only directory copying of the application code (which I cannot add for security reasons) but the issue happens with or without the copying
The newly linked issue could howver be related as we do use the WORKDIR command

`Can you provide an export of the image that fails with /bin/sh not found? docker save will do this for you.`
> Unfortunately no as it contains proprietary code

`What is the run command used to start the containers that fail?`
> The containers are started with a docker-compose up command but even with the basic docker run command I managed to have the container failing on the entrypoint.
I concur it definitely seems to be related to the entrypoint not being copied but I don't see why. There has nto been any changes to the Dockerfile between the upgrade

`Is there anything like AppArmor or SELinux in play? The buildkit build shows a 2B transfer for the build context, which seems to indicate that the entrypoint script was never transferred to the daemon.`
> I did not have Apparmor installed at first, but I did install it when I saw some other users having issues solved with it. Although it didn't change anything.

**Comment by neersighted** [comment]
Ah, `WORKDIR` sounds related. Have you actually successfully reproduced with the example tar you provided? It sounds probable that this is a duplicate of the `WORKDIR` issue and the reproducer you provided doesn't actually reproduce with your instructions due to the missing 

Is the builder cache warm when doing an upgrade, or are you clearing the cache (e.g. did you build on both 20.10 and then 23.0 without clearing the cache first)?

When I ask for a `docker save`, I am asking for **a** failing image. It sounds like you have not actually created a minimal reproducer -- if my above suspicion is correct and you haven't managed to reproduce with the example Dockerfile, can you please modify things until you have a working reproduction & export that image for inspection?

Also, we don't need the entrypoint itself -- just a Dockerfile that reproduces + an export of a failing image ought to be enough to examine.

**Comment by neersighted** [comment]
Also, you say the build looks fine on 23.0, but that is not the case from the sample output you provided. Could you try to more explicitly describe every step in the reproduction chain, along with output?

**Comment by neersighted** [comment]
I've edited your comments for readability as the formatting you used (with the inverted quotes and code blocks) made your comments significantly more challenging to understand.

> No I did not add the workdir command as it did not seem relevant to the issue as the COPY of the entrypoint is made to a completely different (and absolute) path. But I'll try again with.

The first principle of a minimal reproducer is that it must reproduce. Please always make sure to validate something reproduces, in as 'clean' an environment as possible (isolate outside variables and make sure to capture what is necessary to reproduce).

Once you provide a Dockerfile that does reproduce, I think we will find this is in fact a duplicate of #44947. I will likely make that the tracking issue as it has a better reproduction in the issue body.

> I meant that there was no obvious error in the CLI (not counting whatever is in the logs) that would indicate that the entrypoint wasn't copied properly

There is clearly a build failure in the build logs from 23.0. I'm not sure what could be more explicit -- could you please try and provide every **complete** command invocation you use as well as **complete** output so there is no ambiguity?

> I did not enable buildkit with the output of the build I posted earlier

We can only reason in terms of the information you present -- which is why a minimal reproducer is best. If your example steps don't represent the real failures (which in this case, they did not, as it was a BuildKit-less build), there's not much people can do to help.

---

Thanks for being so responsive -- it helps a lot when narrowing things like this down. I hope that you're finding this helpful, even if it is tedious, as well. Please make sure to focus on reproduction with **real, tested** examples and the real output from the exact commands, so those of us with experience working with this code can help you determine the root cause in a timely fashion.

**Comment by rrajkomar** [comment]
Allright, so after tinkering a bit with my builds and googling whatever errors I ended up having.
I finally arrived on this issue which described exactly my issue : https://github.com/docker/buildx/issues/1595
And it turns out that, although it is a bit drastic, running a complete flush of the unused data and cache does solve the issue.

All I can conclude from this is that running the docker build command with buildkit enabled on a docker 23.0 engine when there's remaining data from the 20.10 engine built image somewhere will trigger some issues mainly with RUN commands (even though I did have an issue with an ADD command using a http url).

Once the 20.10 version data is removed, it all works as it was supposed to.
The weird thing is that I was using the --no-cache option and it seems some things were still being used or I wouldn't have had this issue.

Do note though that the issues only happened when buildkit on 23.0 was enabled : if I turned it off, everything worked fine (even without clearing anything)
